### ISSUE TITLE
Set label visibility to certain zoom extents

### ISSUE THREAD

**Issue by nothing-butanol** [tomchadwin/qgis2web#104]
Hi Tom,

I was wondering if it was fairly feasible to add the potential to set the label visibility to certain zooms for leaflet in future releases? It may be more tedious than anything, but the potential enhancement would be really useful for data sets with lots of features.

Thank you for all the work you have done thus far. It's a fantastic plugin.

**Comment by tomchadwin** [comment]
If such scale-dependency is supported in OL3 and the Leaflet label plugin, then this could be possible. If not, it can't be done. Does anyone know?

**Comment by nothing-butanol** [comment]
I was able to add scale dependent labels in Leaflet by modifying bindLabel and adding this in the html file:

        var visible;
        
        map.on('zoomend', function (e) {
            if (map.getZoom() > 13) {
                if (!visible) {
                    json_Aug2015MasterJSON.eachLayer(function (layer) {
                        layer.showLabel();
                    });
                    visible = true;
                }
            } else {
                if (visible) {
                    json_Aug2015MasterJSON.eachLayer(function (layer) {
                        layer.hideLabel();
                    });
                    visible = false;
                }
            }
        });

I found the solution here: http://stackoverflow.com/questions/27820338/leaflet-how-do-i-show-a-label-beyond-a-certain-zoom-level

Sorry for the formatting.

**Comment by tomchadwin** [comment]
Great solution. Is there a way to specify the zoom levels for label visibility in QGIS? If so, we might be able to automate this.

**Comment by riccardoklinger** [comment]
If you save a project they are stored in the qml file. As I remember correctly...

Von meinem iPhone gesendet

**Comment by tomchadwin** [comment]
I'll look into it. Thanks, both.

**Comment by fpuga** [comment]
I'm also interested in this feature. The "scale-based visibility" for labels can be set in Properties -> Labels -> Rendering (I attach a screenshot).

BTW, many thanks for the plugin.

**Comment by tomchadwin** [comment]
Anyone have an OL3 solution for this?

**Comment by dwschmid** [comment]
This might help:
http://stackoverflow.com/questions/33941007/change-labels-opacity-on-a-specific-zoom-inopenlayers

**Comment by tomchadwin** [comment]
Initial Leaflet work done - please test. Note: labels will currently *always* display when the map is initially loaded, regardless of settings - the scale settings will only kick in when you zoom. That's my next thing to fix.

Let me know if it's working (other than initial display). OL3 still to do as well, of course.

**Comment by tomchadwin** [comment]
I can't get the scale-dependent label code to apply on initial page/map load. It works after that whenever you zoom, but the initial display always shows the labels, regardless of label scale-dependent visibility.

Can anyone set up a layer with labels with scale-dependent visibility in QGIS, export to Leaflet, and then see if they can edit the output code to take account of the label setting on initial load? If you can figure it out, I'll add the solution back into the code.

**Comment by tomchadwin** [comment]
Possibly `layer.on("load", function)`:

http://stackoverflow.com/questions/22288933/how-to-tell-when-all-visible-tiles-have-fully-loaded

Will try.

**Comment by tomchadwin** [comment]
Still can't get this to work. Have commented on solution on StackExchange:

http://stackoverflow.com/a/27822424/5613104

**Comment by khiyer** [comment]
Confirm that Leaflet uses the scale dependency but not on the initial display. Couldn't figure out work around.
Zoom dependent labels in OL3 can be achieved by changing the following in xxx_style.js

REPLACE
var labelText = String(feature.get("LABEL"));

WITH
if (resolution < 600) {
   var labelText = String(feature.get("LABEL"));
} else {
   var labelText = ""
}

**Comment by tomchadwin** [comment]
Thanks! In the Leaflet code, can you confirm this works on initial map display? I also don't understand the following lines:

```javascript
layer.hideLabel && layer.hideLabel();
[...]
layer.showLabel && layer.showLabel();
```

In the code I've already implemented, I just do:

```javascript
layer.hideLabel();
[...]
layer.showLabel();
```

**Comment by tomchadwin** [comment]
Ah, I think it might be the `stackLayers()` call which is messing this up. Looking into it now.

**Comment by tomchadwin** [comment]
OK, I think Leaflet might be working now, though I'm slightly concerned I might have left some pointless code in there. Will investigate further, but can you test? Initial display should now work.

**Comment by tomchadwin** [comment]
Yes, I've removed some unused code - please test.

**Comment by khiyer** [comment]
works great. thanks.
Will you be implementing it in OL3 as well?

**Comment by tomchadwin** [comment]
On to OL3. @khiyer, you said to try:

```javascript
if (resolution < 600) {
   var labelText = String(feature.get("LABEL"));
} else {
   var labelText = ""
}
```

I don't know where `resolution` comes from.

**Comment by tomchadwin** [comment]
Looking at it now, yes. Not working yet, but hopefully will be easier than the Leaflet solution.

**Comment by khiyer** [comment]
i'm not sure either as the snippet was forwarded to me. I suspect maybe you can use zoom level or some other variable from OL3 just as well.

**Comment by tomchadwin** [comment]
Sorry, I think yuo have replied to a message I deleted. I found the var, so I'm looking into it now. Not working yet...

**Comment by tomchadwin** [comment]
Now done in OL3 - please test.

**Comment by tomchadwin** [comment]
@khiyer 

> Would it be possible to implement this on a layer-by-layer basis in the GUI

It picks up the values you set in QGIS, so it's already per layer.

**Comment by khiyer** [comment]
ah yes.. forgot about that. tested and confirmed working in OL3 and Leaflet. Thanks.

**Comment by tomchadwin** [comment]
Released as 1.17.0 and pushed to QGIS plugins repo.
